### PR TITLE
Add some possibly useful code ported from https://github.com/huanshankeji/kotlin-common/tree/main/vertx

### PR DESCRIPTION
**Pull request by ShreckYe**
Motivation:

Hi, Vert.x Maintainers. We have been developing our projects in Kotlin with the Vert.x toolkit as the core of the backend. It's been quite smooth and this project provides a lot of useful extensions in Kotlin. During our development, we came up with a few more common extensions in [our shared code repository](https://github.com/huanshankeji/kotlin-common), which we thought could be ported back to this project and made universally available.

We have tried to pick the less opinionated code from our common project in this PR so it's generally useful to most developers using Vert.x in Kotlin. If there is any code that is not appropriate to be added back to this repository, just feel free to remove it or make further changes as it suits your needs. Or if there is none to be added, just close this PR.

Thank you.

### CODE REVIEW

**Comment by vietj** [comment]
can you use vertx-unit instead of vertx-junit5 for the tests for the sake of consistency ?

**Comment by ShreckYe** [comment]
OK, I will try to adapt them.

**Comment by vietj** [comment]
I pinged kotlin users on our discord to give eventually feedback on the PR

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ext/web/ExtendedWebCoroutineVerticle.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package io.vertx.kotlin.coroutines.ext.web
+
+import io.vertx.ext.web.Route
+import io.vertx.ext.web.RoutingContext
+import io.vertx.kotlin.coroutines.CoroutineVerticle
+import io.vertx.kotlin.ext.web.checkedRun
+import kotlinx.coroutines.launch
+
+abstract class ExtendedWebCoroutineVerticle : CoroutineVerticle() {
```
Comment (wowselim) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ext/web/ExtendedWebCoroutineVerticle.kt:9]: Kotlin 1.6.20 added support for [context receivers](https://kotlinlang.org/docs/whatsnew1620.html#prototype-of-context-receivers-for-kotlin-jvm) which IMO is a nicer option compared to a super class.

The functions inside this class could be defined as follows:

```kotlin
context(CoroutineScope)
fun Route.coroutineHandler(handler: suspend (RoutingContext) -> Unit): Route {
  // …
```

This means that they can be used inside the existing `CoroutineVerticle` without having to extend this new verticle. The feature is not yet stable but would avoid having to add a new class that users have to extend.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt [hunk 2]
```diff
@@ -150,9 +160,43 @@ suspend fun <T> Future<T>.await(): T = when {
...
+ * Execute [blockingCode] that returns the a [T] instance with [Vertx.executeBlocking]
+ * and awaits its completion.
+ *
+ * Compared to [Vertx.executeBlocking]'s `blockingCodeHandler` argument,
+ * [blockingCode] returns the result when the operation completes instead of calling [Promise.complete].
+ */
+suspend fun <T> Vertx.awaitExecuteBlocking(blockingCode: () -> T): T =
+  executeBlocking<T> {
+    it.complete(blockingCode())
+  }.await()
```
Comment (wowselim) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:188]: How is this different from the existing `io.vertx.kotlin.coroutines.VertxCoroutineKt#awaitBlocking`?

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt [hunk 3]
```diff
@@ -150,9 +160,43 @@ suspend fun <T> Future<T>.await(): T = when {
...
+
+/**
+ * Like [awaitExecuteBlocking] but [blockingCode] is a suspend function.
+ */
+suspend fun <T> Vertx.awaitSuspendExecuteBlocking(blockingCode: suspend () -> T): T =
+  coroutineScope {
+    executeBlocking<T> {
+      launch { it.complete(blockingCode()) }
+    }.await()
+  }
```
Comment (wowselim) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:198]: Do you have an example of when this function would be useful?

**Comment by wowselim** [comment]
There's some pretty neat additions proposed here that should help reduce boilerplate in most kotlin projects 👍

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt [hunk 4]
```diff
@@ -15,23 +15,13 @@
  */
 package io.vertx.kotlin.coroutines
 
-import io.vertx.core.AsyncResult
-import io.vertx.core.Context
+import io.vertx.core.*
```
Comment (jojopad) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:18]: I would suggest avoiding using the star import and instead use a specific import statement for each class used.

File: vertx-lang-kotlin-coroutines/src/test/kotlin/io/vertx/kotlin/coroutines/VertxCoroutineTest.kt [hunk 5]
```diff
@@ -15,32 +15,24 @@
  */
 package io.vertx.kotlin.coroutines
 
-import io.vertx.core.Context
-import io.vertx.core.Future
-import io.vertx.core.Promise
-import io.vertx.core.Vertx
+import io.vertx.core.*
```
Comment (jojopad) [inline vertx-lang-kotlin-coroutines/src/test/kotlin/io/vertx/kotlin/coroutines/VertxCoroutineTest.kt:18]: How about retaining the previous import statements which don't use the star import?

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ext/web/ExtendedWebCoroutineVerticle.kt (hunk 1 above)
Comment (jojopad) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ext/web/ExtendedWebCoroutineVerticle.kt:9]: While the context receivers seems to be a cleaner option, it's not yet recommended for production use at least for the latest Kotlin 1.6.20 release.

**Comment by jojopad** [comment]
Some updates with the [documentation and code samples](https://github.com/vert-x3/vertx-lang-kotlin/blob/master/vertx-lang-kotlin/src/main/asciidoc/vertx-core/kotlin/index.adoc) how to use the proposed new vertx kotlin extensions would be really helpful

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ext/web/ExtendedWebCoroutineVerticle.kt (hunk 1 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ext/web/ExtendedWebCoroutineVerticle.kt:9]: Yeah, I agree. It's just still in prototype.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 2 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:188]: Sorry, I forgot about such a function. It's basically the same.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 3 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:198]: No, not yet, but I think since `executeBlocking` offers a `Promise` to produce the result asynchronously such a use case exists.

**Comment by ShreckYe** [comment]
> There's some pretty neat additions proposed here that should help reduce boilerplate in most kotlin projects +1

Thank you.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 4 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:18]: True. IntelliJ IDEA likes to format with start imports and I don't know how to disable that.

File: vertx-lang-kotlin-coroutines/src/test/kotlin/io/vertx/kotlin/coroutines/VertxCoroutineTest.kt (hunk 5 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/test/kotlin/io/vertx/kotlin/coroutines/VertxCoroutineTest.kt:18]: No problem. I will sort them out then.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 4 above)
Comment (jojopad) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:18]: This tip worked for me https://stackoverflow.com/questions/48755419/disable-intellij-kotlin-starred-imports

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 2 above)
Comment (wowselim) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:188]: There's also a test for `awaitBlocking` (`io.vertx.kotlin.coroutines.VertxCoroutineTest#test awaitBlocking`). I still don't see the difference between the two functions…

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 3 above)
Comment (wowselim) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:198]: I think this is already possible with awaitBlocking
```kotlin
vertx.awaitSuspendExecuteBlocking {
  val blockingCode = awaitBlocking { Files.read(...) }
}
```

There's no need to run this inside `executeBlocking` because `awaitBlocking` suspends the coroutine and therefore I don't see why this lambda should run on a worker thread. Am I missing something?

Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:198]: OK I have come up with an example:

```kotlin
// In real-world scenarios this could be hashing, compressing, encrypting, etc.
fun mockSomeExpensiveComputation(): Int {
  Thread.sleep(3000)
  return Random.nextInt()
}

suspend fun mockSomeExpensiveComputationWithAsyncIO() {
  val asyncFile = Vertx.currentContext().owner().fileSystem()
    .open("test_file", OpenOptions().setWrite(true)).await()
  repeat(2) { i ->
    asyncFile.write(Buffer.buffer(4).appendInt(mockSomeExpensiveComputation()))
  }
}

fun main() {
  Vertx.vertx().run {
    CoroutineScope(dispatcher()).launch {
      println("On event loop thread.")
      mockSomeExpensiveComputationWithAsyncIO()
      awaitSuspendExecuteBlocking {
        println("On worker thread.")
        mockSomeExpensiveComputation()
        println("Done.")
      }
    }
  }
}
```

Suppose that there is a `suspend` function, for example `mockSomeExpensiveComputationWithAsyncIO` in the code above, which has blocking computations interleaved with asynchronous suspending IO operations. And suppose that this is a library function that is not easy to for you to edit, so you can't wrap every `mockSomeExpensiveComputation` with `awaitBlocking`. This is when `awaitSuspendExecuteBlocking` can come into use.

BTW, to make it consistent with the `awaitBlocking`, I think this function can be renamed into `awaitBlockingSuspend` or `awaitSuspendBlocking`.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 4 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:18]: It works. Thank you.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 3 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:198]: There seems to be a bug that the coroutine is launched using the event loop thread as its dispatcher. I will try to fix this.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 2 above)
Comment (wowselim) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:188]: @ShreckYe yes, the exceptions are different because their stack traces differ.

Let me rephrase: When would you use this function over `awaitBlocking`? 😄

Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:188]: There is almost no difference. I added this because I didn't know there was already an `awaitBlocking` function. so I think I will remove this one in my next commit.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 3 above)
Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt:198]: What do you think is a good name for this? The ones I can think of are:
1. awaitBlockingSuspend
1. awaitBlockingCoroutine
1. awaitSuspendBlocking
1. awaitCoroutineBlocking

**Comment by ShreckYe** [comment]
I don't fully understand why the CI tests failed. Is it because that the "kotlin" source set in "vertx-lang-kotlin" is overriden by generated code?

**Comment by vietj** [comment]
do we have consensus on this PR ?

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt (hunk 3 above)
Comment (wowselim) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt]: Which thread is this lambda going to run on?

```kotlin
launch { it.complete(blockingCode()) }
```

Comment (ShreckYe) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt]: This implementation is buggy and it runs on the event loop thread. I fixed it [here](https://github.com/vert-x3/vertx-lang-kotlin/pull/222/files#diff-162b76dc534138518a237d9a8ed527f1b3ecaca67385ea7d4357b6eff203f699R138-R160) and [the test](https://github.com/vert-x3/vertx-lang-kotlin/pull/222/files#diff-b4162a5ef737adccaa47e170e53212572f2c942a94dc94bbcc166555ea5e2699R409-R413) shows that the fixed version runs on the worker thread.

Comment (wowselim) [reply vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt]: nice 👍🙂

**Comment by vietj** [comment]
I don't know what is wrong with tests not passing, have you tried to build locally ?

**Comment by ShreckYe** [comment]
The Maven `test` phases pass locally. I moved Tuple.kt to the `java` source directory. Maybe this will work.

**Comment by tsegismont** [comment]
Superseded by https://github.com/vert-x3/vertx-lang-kotlin/pull/253
